src/Func.cpp

```diff
@@ -602,529 +608,424 @@ class SubstituteSelfReference : public IRMutator {
 
-        return true;
+    // Check that no Vars were fused into RVars
+    Scope<> rdims;
+    for (const ReductionVariable &rv : definition.schedule().rvars()) {
+        rdims.push(rv.var);
+    }
+    for (const Split &split : definition.schedule().splits()) {
+        switch (split.split_type) {
+        case Split::SplitVar:
+            if (rdims.contains(split.old_var)) {
+                rdims.pop(split.old_var);
+                rdims.push(split.outer);
+                rdims.push(split.inner);
+            }
+            break;
+        case Split::FuseVars:
+            if (rdims.contains(split.outer) || rdims.contains(split.inner)) {
+                if (!(rdims.contains(split.outer) && rdims.contains(split.inner))) {
+                    std::stringstream s;
+                    s << "Cannot rfactor a Func that has fused a Var into an RVar ("
+                      << split.outer << ", " << split.inner << ")";
+                    return s.str();
+                }
+                rdims.pop(split.outer);
+                rdims.pop(split.inner);
+                rdims.push(split.old_var);
+            }
+            break;
+        case Split::PurifyRVar:
+        case Split::RenameVar:
+            if (rdims.contains(split.old_var)) {
+                rdims.pop(split.old_var);
+                rdims.push(split.outer);
+            }
+            break;
+        }
     }
-    return false;
-}
 
-/** Apply purify directives on the reduction variables and predicates. Purify
- * replace a RVar with a Var, thus, the RVar needs to be removed from the list.
- * Any reference to the RVar in the predicates will be replaced with reference
- * to a Var. */
-bool apply_purify(const Split &s, vector<ReductionVariable> &rvars,
-                  vector<Expr> &predicates, vector<Expr> &args,
-                  vector<Expr> &values, map<string, Expr> &dim_extent_alignment) {
-    internal_assert(s.split_type == Split::PurifyRVar);
-    const auto &iter = std::find_if(rvars.begin(), rvars.end(),
-                                    [&s](const ReductionVariable &rv) { return (s.old_var == rv.var); });
-    if (iter != rvars.end()) {
-        debug(4) << "  Purify RVar " << iter->var << " into Var " << s.outer
-                 << ", deleting it from the rvars list\n";
-        rvars.erase(iter);
+    return std::nullopt;
+}
 
-        vector<ApplySplitResult> splits_result = apply_split(s, "", dim_extent_alignment);
-        vector<pair<string, Expr>> bounds_let_stmts = compute_loop_bounds_after_split(s, "");
-        apply_split_result(bounds_let_stmts, splits_result, predicates, args, values);
+using SubstitutionMap = std::map<string, Expr>;
 
-        return true;
+// This is a helper function for building up a substitution map that
+// corresponds to pushing down a nest of lets. The lets should be fed
+// to this function from innermost to outermost. This is equivalent to
+// building a let-nest as a one-hole context and then simplifying.
```

Comment:
What is a "one-hole context"?

Reply:
It's an AST where one of the nodes is missing a single child. In this case, the innermost let is missing its body.

For example,

```
let x = 30
in let y = 2*x+z
   in (hole)
```
Now we can fill that hole with something like `y + 3` and when we simplify, we get `63 + z`.